substrate/frame/assets/src/functions.rs

```diff
@@ -654,11 +688,20 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 			debug_assert!(source_account.balance >= debit, "checked in prep; qed");
 			source_account.balance = source_account.balance.saturating_sub(debit);
 
+			// Pre-check that an account can die if is below min balance
+			if source_account.balance < details.min_balance {
+				source_died =
+					Some(Self::dead_account(source, details, &source_account.reason, false));
```

Comment:
it does mutate the state, should be called ones

Reply:
In this specific case, we're safe.

1. Since the block for `Asset::try_mutate` hasn't finished yet, failing the ensure at this point would cause no mutations for `details`.
2. Also, no `dest` account mutations happened here because we're doing this before the `Account::try_mutate` block.
3. Finally, mutations for `source_account` only get saved at the end, much later than this step (see lines (735, 739).

Reply:
`fn dead_account` modifies the storage `frame_system::Account` and potentially anything.
Those are not reverted AFAICT

Reply:
Let me write a couple of additional tests to assert something, because from what I've seen in previous changes I've done in this sense is that if you run these ensures before knowing the account might die is that you get false positives (the call fails, but it was never dying).

Reply:
Done ([`3291c66`](https://github.com/paritytech/polkadot-sdk/pull/4530/commits/3291c66e62d0813270dd0f53bcb5998412db08c7)).

Final note here: `transfer_and_die` indirectly calls `can_decrease` (via `prep_debit`). Thus, it's nearly impossible that the account has passed through the already tested \[[1](https://github.com/paritytech/polkadot-sdk/blob/3291c66e62d0813270dd0f53bcb5998412db08c7/substrate/frame/assets/src/functions.rs#L302)\] minimum balance checks.

Reply:
indeed I see the check `ensure_account_can_die` is redundant with prep_debit, but maybe it is good to keep.